File: frontend/lib/models/user_role.dart
/// Rôles utilisateur. Valeurs API alignées sur `RoleNames` côté back.
enum UserRole {
  technician('Technician', 'Technicien'),
  admin('Admin', 'Administrateur'),
  supervisor('Supervisor', 'Superviseur');

  const UserRole(this.apiValue, this.label);

  /// Valeur stockée en base et émise dans le JWT.
  final String apiValue;

  /// Libellé affiché dans l'interface.
  final String label;

  /// Rôle reconnu, ou `null` pour une valeur inconnue. Pas de repli sur
  /// technicien : un rôle non géré par cette version de l'app ne doit pas
  /// ouvrir l'espace de création de CRI (cf. docs/plan-role-superviseur.md §3).
  static UserRole? fromString(String? role) {
    switch (role?.trim().toLowerCase()) {
      case 'technician':
      case 'technicien':
        return UserRole.technician;
      case 'admin':
        return UserRole.admin;
      case 'supervisor':
        return UserRole.supervisor;
      default:
        return null;
    }
  }
}

File: backend/src/NovadisApi/Models/UserRole.cs
namespace NovadisApi.Models
{
    /// <summary>
    /// Énumération des rôles utilisateur
    /// </summary>
    public enum UserRole
    {
        Technician,
        Admin,
        Supervisor
    }

    /// <summary>
    /// Valeurs canoniques des rôles, telles que stockées en base et émises dans le JWT.
    /// </summary>
    public static class RoleNames
    {
        public const string Technician = "Technician";
        public const string Admin = "Admin";
        public const string Supervisor = "Supervisor";
    }

    /// <summary>
    /// Extensions pour le rôle utilisateur
    /// </summary>
    public static class UserRoleExtensions
    {
        /// <summary>
        /// Rôle reconnu, ou <c>null</c> pour une valeur inconnue. Pas de repli sur
        /// Technician : un rôle mal saisi en base donnerait sinon le droit de créer
        /// des CRI (cf. docs/plan-role-superviseur.md §3).
        /// </summary>
        public static UserRole? FromString(string? role)
        {
            return role?.Trim().ToLowerInvariant() switch
            {
                "admin" => UserRole.Admin,
                "technician" => UserRole.Technician,
                "technicien" => UserRole.Technician,
                "supervisor" => UserRole.Supervisor,
                _ => null
            };
        }

        public static string ToRoleString(this UserRole role)
        {
            return role switch
            {
                UserRole.Admin => RoleNames.Admin,
                UserRole.Supervisor => RoleNames.Supervisor,
                _ => RoleNames.Technician
            };
        }

        /// <summary>
        /// Forme canonique d'un rôle stocké ("Technicien" → "Technician").
        /// À utiliser partout où le rôle sort de la base (JWT, DTO) : les policies
        /// ne connaissent que les valeurs de <see cref="RoleNames"/>. Un rôle
        /// inconnu est renvoyé tel quel — il ne correspond à aucune capacité.
        /// </summary>
        public static string Normalize(string role) => FromString(role)?.ToRoleString() ?? role;
    }
}
